cpp/src/io/orc/reader.cu

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "io/orc/reader_impl.hpp"
+#include "io/orc/reader_impl_helpers.hpp"
+
+namespace cudf::io::orc::detail {
+
+// Destructor are defined within this translation unit.
+reader::~reader() = default;
+
+reader::reader(std::vector<std::unique_ptr<cudf::io::datasource>>&& sources,
+               orc_reader_options const& options,
+               rmm::cuda_stream_view stream,
+               rmm::mr::device_memory_resource* mr)
+  : _impl{std::make_unique<reader_impl>(std::move(sources), options, stream, mr)}
+{
+}
+
+table_with_metadata reader::read() { return _impl->read(); }
+
+chunked_reader::chunked_reader(std::size_t output_size_limit,
+                               std::size_t data_read_limit,
+                               std::vector<std::unique_ptr<datasource>>&& sources,
+                               orc_reader_options const& options,
+                               rmm::cuda_stream_view stream,
+                               rmm::mr::device_memory_resource* mr)
+  : _impl{std::make_unique<reader_impl>(
+      output_size_limit, data_read_limit, std::move(sources), options, stream, mr)}
+{
+}
+
+chunked_reader::chunked_reader(std::size_t output_size_limit,
+                               std::size_t data_read_limit,
+                               size_type output_row_granularity,
+                               std::vector<std::unique_ptr<datasource>>&& sources,
+                               orc_reader_options const& options,
+                               rmm::cuda_stream_view stream,
+                               rmm::mr::device_memory_resource* mr)
+  : _impl{std::make_unique<reader_impl>(output_size_limit,
+                                        data_read_limit,
+                                        output_row_granularity,
+                                        std::move(sources),
+                                        options,
+                                        stream,
+                                        mr)}
+{
+  // Although we internally accept non-positive value for `output_row_granularity` because we
+  // implicitly change such value into `DEFAULT_OUTPUT_ROW_GRANULARITY`.
+  // The user are not allowed to do so but instead required to specify an explicit positive number.
```

Comment:
```suggestion
  // Users are not allowed to do so but instead required to specify an explicit positive number.
```

Reply:
wait, but they are allowed to pass anything <= 0. It will silently become DEFAULT_OUTPUT_ROW_GRANULARITY.
I think we should throw for non-positive numbers instead of replacing the value. It's quite literally an invalid parameter.

Reply:
Users are not allowed to pass any number `<=0` to the **public interface**. However, we (dev) are allowed, and we do that through our internal detail interface.

Humnn, maybe that is a bit confusing here. I can change that somehow to totally eliminate it.

Reply:
Done.